File: scripts/dev_preflight.py
#!/usr/bin/env python3
"""Non-interactive dev preflight for `make dev`.

Reports each dependency the chosen run mode needs, with an actionable fix, and
exits non-zero when a hard blocker is missing so `make dev` can stop and point
the developer at the fix (req 3: no silent half-starts). It reuses murmur's own
music preflight (spec 03-03) so this and the in-app startup check agree.

Usage:
    python scripts/dev_preflight.py --voice spark   # real TTS wanted
    python scripts/dev_preflight.py --voice stub    # silent voice (no mlx/model)
    python scripts/dev_preflight.py --no-music      # skip the binary checks
"""

from __future__ import annotations

import argparse
import asyncio
import os
import platform
import sys

from murmur.music.preflight import preflight_ffmpeg, preflight_ytdlp

_OK = "\033[32m✓\033[0m"
_NO = "\033[31m✗\033[0m"


async def _probe_music() -> list[tuple[str, bool, str]]:
    """(name, ok, reason) for each music binary, probed concurrently."""
    yt, ff = await asyncio.gather(preflight_ytdlp(), preflight_ffmpeg())
    return [("yt-dlp", yt.ok, yt.reason), ("ffmpeg", ff.ok, ff.reason)]


def _check_music() -> list[str]:
    """Print each binary's status; return fix hints for the broken ones."""
    problems: list[str] = []
    for name, ok, reason in asyncio.run(_probe_music()):
        if ok:
            print(f"  {_OK} {name}")
        else:
            print(f"  {_NO} {name}: {reason}")
            problems.append(name)
    if problems:
        joined = " ".join(problems)
        return [
            f"music needs {joined}. Fix with:  make setup-music"
            f"   (or: brew install {joined})",
        ]
    return []


def _check_real_voice() -> list[str]:
    """Real MLX voices are Apple-Silicon only and need the tts-mlx extra."""
    on_apple_silicon = platform.system() == "Darwin" and platform.machine() == "arm64"
    if not on_apple_silicon:
        print(f"  {_NO} real voice: needs Apple Silicon (this host is not)")
        return ["real voice unavailable here. Use:  VOICE=stub make dev"]
    try:
        import mlx_audio  # noqa: F401  (probe only)
    except ImportError:
        print(f"  {_NO} real voice: mlx-audio not installed")
        return ["install the TTS extra:  uv sync --extra tts-mlx"]
    print(f"  {_OK} real voice (mlx-audio installed)")
    print("      note: the first run downloads the TTS model (slow, once).")
    return []


def _check_remote_voice() -> list[str]:
    """The remote HTTP backend (spec 02 §3.6) needs no local model — just a
    configured endpoint. Check the URL is set; the server itself is proven on the
    first synth."""
    url = os.environ.get("MURMUR_TTS_URL", "")
    if not url:
        print(f"  {_NO} remote voice: MURMUR_TTS_URL not set")
        return [
            "remote voice needs an endpoint:  "
            "MURMUR_TTS_URL=http://host:port VOICE=remote make dev"
        ]
    print(f"  {_OK} remote voice -> {url}")
    return []


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="dev preflight for make dev")
    ap.add_argument("--voice", default="spark", help="voice backend (or 'stub')")
    ap.add_argument("--no-music", action="store_true", help="skip music binaries")
    args = ap.parse_args(argv)

    print("preflight:")
    fixes: list[str] = []
    if not args.no_music:
        fixes += _check_music()
    if args.voice == "remote":
        fixes += _check_remote_voice()  # off-machine HTTP TTS — no local model
    elif args.voice != "stub":
        fixes += _check_real_voice()

    if fixes:
        print("\nblockers — fix these, or use an escape hatch:")
        for fix in fixes:
            print(f"  → {fix}")
        print("  → skip everything (offline):  STUB=1 make dev")
        return 1
    print("all set.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/dev-preflight.ts
// Non-interactive dev preflight for `make dev`.
//
// A REPORTER, not a gatekeeper (spec 03-03 §7.1): it names each dependency the
// chosen run mode wants and what murmur will do without it, then exits 0 —
// always. murmur assumes the user has Claude Code, so every gap named here is
// one the app repairs by TALKING to the user (the setup conversation offered at
// startup, or `make setup`), which means the shell must never stand between
// them and the app. The one exception needs no code: without `node` the shell
// cannot run this script at all, and `make dev` stops there (§7.3 criterion 8).
//
//   node scripts/dev-preflight.ts --voice hosted   # remote TTS wanted
//   node scripts/dev-preflight.ts --voice stub     # silent voice
//   node scripts/dev-preflight.ts --no-music       # skip the binary checks
//   node scripts/dev-preflight.ts --plain          # plain front-end, no bun

import { parseArgs } from 'node:util'

import { voiceConfigPath } from '../src/paths.ts'
import {
  preflightBun,
  preflightFfmpeg,
  preflightYtdlp,
  preflightYtdlpFreshness,
} from '../src/setup/startup.ts'
import { readVoiceConfig } from '../src/voice/voice-config.ts'

const OK = '\x1b[32m✓\x1b[0m'
const NO = '\x1b[33m·\x1b[0m'

const { values } = parseArgs({
  args: process.argv.slice(2),
  options: {
    voice: { type: 'string' },
    'no-music': { type: 'boolean' },
    plain: { type: 'boolean' },
    tui: { type: 'boolean' },
  },
})
const voice = values.voice ?? 'hosted'
// The TUI is the default front-end (spec 10 §6); --plain is the opt-out.
const wantsTui = values.plain !== true

console.log('preflight:')
// What the session will be missing, in the words the app itself will use.
const gaps: string[] = []

if (values['no-music'] !== true) {
  const [yt, ff, fresh] = await Promise.all([
    preflightYtdlp(),
    preflightFfmpeg(),
    preflightYtdlpFreshness(),
  ])
  const broken: string[] = []
  for (const [name, result] of [
    ['yt-dlp', yt],
    ['ffmpeg', ff],
  ] as const) {
    if (result.ok) console.log(`  ${OK} ${name}`)
    else {
      console.log(`  ${NO} ${name}: ${result.reason}`)
      broken.push(name)
    }
  }
  if (broken.length > 0) gaps.push(`music needs ${broken.join(' + ')} — the session will be talk-only`)
  else if (!fresh.ok) {
    // The binaries work — this is a freshness warning, not a degradation:
    // music keeps playing, but extractors rot (Bilibili breaks first).
    console.log(`  ${NO} yt-dlp freshness: ${fresh.reason}`)
    gaps.push('yt-dlp is stale (Bilibili breaks first as sites change) — an upgrade fixes it')
  }
}

if (wantsTui) {
  const bun = await preflightBun()
  if (bun.ok) console.log(`  ${OK} bun (tui front-end)`)
  else {
    console.log(`  ${NO} bun: ${bun.reason}`)
    gaps.push('the tui front-end needs bun — the plain front-end will be used')
  }
}

if (voice === 'hosted') {
  // The endpoint is proven on the first synth; here we only report whether one
  // is configured at all, from either layer the app reads (spec 03-03 §7.2).
  const url = process.env.MURMUR_TTS_URL?.trim() || (readVoiceConfig(voiceConfigPath())?.ttsUrl ?? '')
  if (url === '') {
    console.log(`  ${NO} voice endpoint: not configured yet`)
    gaps.push('the voice has no endpoint — lines will show as text, silently')
  } else {
    console.log(`  ${OK} hosted voice -> ${url}`)
  }
}

if (gaps.length > 0) {
  console.log('\nmurmur will start anyway, with these gaps:')
  for (const gap of gaps) console.log(`  · ${gap}`)
  console.log('\nit will offer to fix them by talking you through it. You can also run:')
  console.log('  make setup        the whole surface, as a conversation')
  console.log('  make setup-music  just the music binaries')
} else {
  console.log('all set.')
}
